src/main/java/com/uber/cadence/testing/WorkflowShadower.java

```diff
@@ -0,0 +1,116 @@
+
+public final class WorkflowShadower {
+
+  private final ShadowingOptions options;
+  private final String query;
+  private final ScanWorkflowActivity scanWorkflow;
+  private final ReplayWorkflowActivity replayWorkflow;
+
+  public WorkflowShadower(IWorkflowService service, ShadowingOptions options, String taskList) {
+
+    this(service, options, taskList, new NoopScope());
+  }
+
+  public WorkflowShadower(
+      IWorkflowService service, ShadowingOptions options, String taskList, Scope metricsScope) {
+    this(
+        options,
+        new ScanWorkflowActivityImpl(service),
+        new ReplayWorkflowActivityImpl(service, metricsScope, taskList));
+  }
+
+  public WorkflowShadower(
+      ShadowingOptions options,
+      ScanWorkflowActivity scanWorkflow,
+      ReplayWorkflowActivity replayWorkflow) {
+    this.options = options;
+    this.query = options.getWorkflowQuery();
+    this.scanWorkflow = scanWorkflow;
+    this.replayWorkflow = replayWorkflow;
+  }
+
+  public void run() throws Throwable {
+    byte[] nextPageToken = null;
+    int replayCount = 0;
+
+    int maxReplayCount = Integer.MAX_VALUE;
+    Duration maxReplayDuration = Duration.ZERO;
+    ZonedDateTime now = ZonedDateTime.now();
+    if (options.getExitCondition() != null) {
+      if (options.getExitCondition().getShadowCount() != 0) {
+        maxReplayCount = options.getExitCondition().getShadowCount();
+      }
+      if (options.getExitCondition().getExpirationIntervalInSeconds() != 0) {
+        maxReplayDuration =
+            Duration.ofSeconds(options.getExitCondition().getExpirationIntervalInSeconds());
+      }
+    }
+    do {
+      ScanWorkflowActivityResult ScanResult =
+          scanWorkflow.scan(
+              new ScanWorkflowActivityParams()
+                  .setDomain(options.getDomain())
+                  .setWorkflowQuery(query)
+                  .setNextPageToken(nextPageToken)
+                  .setSamplingRate(options.getSamplingRate()));
+      nextPageToken = ScanResult.getNextPageToken();
+
+      for (WorkflowExecution execution : ScanResult.getExecutions()) {
+        ReplayWorkflowActivityResult replayResult =
+            replayWorkflow.replay(
+                new ReplayWorkflowActivityParams()
+                    .setDomain(options.getDomain())
+                    .setExecutions(Lists.newArrayList(execution)));
+
+        if (replayResult.getFailed() > 0) {
+          throw new Error("Replay workflow history failed with execution:" + execution.toString());
+        } else if (replayResult.getSucceeded() > 0) {
+          replayCount++;
+        }
+
+        // Check exit condition
+        if (replayCount >= maxReplayCount) {
+          return;
+        }
+        if (!maxReplayDuration.isZero()
+            && ZonedDateTime.now().isAfter(now.plusSeconds(maxReplayDuration.getSeconds()))) {
+          return;
+        }
+      }
+    } while (nextPageToken != null && options.getShadowMode() == Mode.Normal);
```

Comment:
add wait period between two scan iterations?

Reply:
Added

Reply:
hmmm, looks like this sleep interval will also be applied with in an iteration, which is not what we want?

Reply:
Good point. Updated